server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java

```diff
@@ -42,7 +42,7 @@ public void refreshShard(
             case WAIT_UNTIL -> waitUntil(indexShard, location, new ActionListener<>() {
                 @Override
                 public void onResponse(Boolean forced) {
-                    if (indexShard.getReplicationGroup().getRoutingTable().unpromotableShards().size() > 0) {
+                    if (indexShard.getReplicationGroup().getRoutingTable().unpromotableShards().size() > 0 && location != null) {
```

Comment:
It looks like we have null location coming from following places: 
https://github.com/elastic/elasticsearch/blob/4b7f87b25dab3e8cd187e56bb5a847874bd3869f/server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseSyncAction.java#L166

https://github.com/elastic/elasticsearch/blob/4b7f87b25dab3e8cd187e56bb5a847874bd3869f/server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseSyncAction.java#L177

https://github.com/elastic/elasticsearch/blob/4b7f87b25dab3e8cd187e56bb5a847874bd3869f/server/src/main/java/org/elasticsearch/action/resync/TransportResyncReplicationAction.java#L114-L121

but I have no understanding about the reason and implications for us

Reply:
Ok. I was confused and distracted by something else. Anything along the lines of:
```java
client().prepareBulk().add(client().prepareUpdate("test", "1").setDoc(Requests.INDEX_CONTENT_TYPE, "field", "2")).setRefreshPolicy(WAIT_UNTIL).get();
```
would return an NPE error in the response instead of missing_document_exception.